components/datetime/src/fields.rs

```diff
@@ -0,0 +1,260 @@
+pub enum FieldSymbol {
+    Era,
+    Year(Year),
+    Month(Month),
+    Day(Day),
+    Weekday(Weekday),
+    Period(Period),
+    Hour(Hour),
+    Minute,
+    Second(Second),
+}
+
+impl FieldSymbol {
+    pub fn from_byte(b: u8) -> Result<Self, Error> {
+        match b {
+            b'G' => Ok(Self::Era),
+            b'm' => Ok(Self::Minute),
+            _ => Year::from_byte(b)
+                .map(Self::Year)
+                .or_else(|_| Month::from_byte(b).map(Self::Month))
+                .or_else(|_| Day::from_byte(b).map(Self::Day))
+                .or_else(|_| Weekday::from_byte(b).map(Self::Weekday))
+                .or_else(|_| Period::from_byte(b).map(Self::Period))
+                .or_else(|_| Hour::from_byte(b).map(Self::Hour))
+                .or_else(|_| Second::from_byte(b).map(Self::Second)),
+        }
+    }
+}
+
+#[allow(dead_code)]
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum Year {
+    Calendar,
+    WeekOf,
+    Extended,
+    Cyclic,
+    RelatedGregorian,
+}
+
+impl Year {
+    pub fn from_byte(b: u8) -> Result<Self, Error> {
+        match b {
+            b'y' => Ok(Self::Calendar),
+            b'Y' => Ok(Self::WeekOf),
+            _ => Err(Error::Unknown),
+        }
+    }
+}
+
+impl From<Year> for FieldSymbol {
+    fn from(input: Year) -> Self {
+        FieldSymbol::Year(input)
+    }
+}
+
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum Month {
+    Format,
+    StandAlone,
+}
+
+impl Month {
+    pub fn from_byte(b: u8) -> Result<Self, Error> {
+        match b {
+            b'M' => Ok(Self::Format),
+            b'L' => Ok(Self::StandAlone),
+            _ => Err(Error::Unknown),
+        }
+    }
+}
+
+impl From<Month> for FieldSymbol {
+    fn from(input: Month) -> Self {
+        FieldSymbol::Month(input)
+    }
+}
+
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum Day {
+    DayOfMonth,
+    DayOfYear,
+    DayOfWeekInMonth,
+    ModifiedJulianDay,
+}
+
+impl Day {
+    pub fn from_byte(b: u8) -> Result<Self, Error> {
+        match b {
+            b'd' => Ok(Self::DayOfMonth),
+            b'D' => Ok(Self::DayOfYear),
+            b'F' => Ok(Self::DayOfWeekInMonth),
+            b'g' => Ok(Self::ModifiedJulianDay),
+            _ => Err(Error::Unknown),
+        }
+    }
+}
+
+impl From<Day> for FieldSymbol {
+    fn from(input: Day) -> Self {
+        FieldSymbol::Day(input)
+    }
+}
+
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum Hour {
+    H11,
+    H12,
+    H23,
+    H24,
+    Preferred,
+    PreferredNoDayPeriod,
+    PreferredFlexibleDayPeriod,
+}
+
+impl Hour {
+    pub fn from_byte(b: u8) -> Result<Self, Error> {
+        match b {
+            b'K' => Ok(Self::H11),
+            b'h' => Ok(Self::H12),
+            b'H' => Ok(Self::H23),
+            b'k' => Ok(Self::H24),
+            b'j' => Ok(Self::Preferred),
+            b'J' => Ok(Self::PreferredNoDayPeriod),
+            b'C' => Ok(Self::PreferredFlexibleDayPeriod),
```

Comment:
`j`, `J`, and `C` don't occur in patterns.  According to [UTS 35](https://unicode.org/reports/tr35/tr35-dates.html#dfst-hour):

> ***Input skeleton symbol***
> It must not occur in pattern or skeleton data. Instead, it is reserved for use in skeletons passed to APIs doing flexible date pattern generation.